File: app/src/main/java/awais/instagrabber/utils/KeywordsFilterUtils.java
package awais.instagrabber.utils;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import awais.instagrabber.repositories.responses.Caption;
import awais.instagrabber.repositories.responses.Media;

public final class KeywordsFilterUtils {

    private final ArrayList<String> keywords;

    public KeywordsFilterUtils(ArrayList<String> keywords){
        this.keywords = keywords;
    }

    public boolean filter(final String caption){
        if(caption == null) return false;
        final String temp = caption.toLowerCase(Locale.getDefault());
        for(final String s:keywords){
            if(temp.contains(s)) return true;
        }
        return false;
    }

    public boolean filter(final Media media){
        if(media == null) return false;
        final Caption c = media.getCaption();
        if(c == null) return false;
        final String temp = c.getText().toLowerCase(LocaleUtils.getCurrentLocale());
        for(final String s:keywords){
            if(temp.contains(s)) return true;
        }
        return false;
    }

    public List<Media> filter(final List<Media> media){
        if(media == null) return new ArrayList<>();

        final List<Media> result= new ArrayList<>();
        for(final Media m:media){
            if(filter(m)) result.add(m);
        }
        return result;
    }
}

File: app/src/main/java/awais/instagrabber/utils/KeywordsFilterUtils.kt
package awais.instagrabber.utils

import awais.instagrabber.repositories.responses.Media
import java.util.*
import kotlin.collections.ArrayList

//    fun filter(caption: String?): Boolean {
//        if (caption == null) return false
//        if (keywords.isEmpty()) return false
//        val temp = caption.toLowerCase()
//        for (s in keywords) {
//            if (temp.contains(s)) return true
//        }
//        return false
//    }

private fun containsAnyKeyword(keywords: List<String>, media: Media?): Boolean {
    if (media == null || keywords.isEmpty()) return false
    val (_, text) = media.caption ?: return false
    val temp = text!!.lowercase(Locale.getDefault())
    return keywords.any { temp.contains(it) }
}

fun filter(keywords: List<String>, media: List<Media>?): List<Media>? {
    if (keywords.isEmpty()) return media
    if (media == null) return ArrayList()
    val result: MutableList<Media> = ArrayList()
    media.filterNotTo(result) { containsAnyKeyword(keywords, it) }
    return result
}